components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt

```diff
@@ -24,27 +26,105 @@ class DefaultLoginValidationDelegate(
     private val scope: CoroutineScope = CoroutineScope(IO)
 ) : LoginValidationDelegate {
 
-    @Suppress("ComplexMethod") // This method is not actually complex
-    override fun validateCanPersist(login: Login): Deferred<Result> {
+    /**
+     * Uses saved [Login]s and new [Login] to return a [Result] for knowing whether an existing
+     * login record can and should be updated or created.
+     */
+    @Suppress("ComplexMethod")
+    override fun getNewLoginPersistenceAsync(newLogin: Login): Deferred<Result> {
         return scope.async {
-            try {
-                // We're setting guid=null here in order to trigger de-duping logic properly.
-                // Internally, the library ensures to not dupe records against themselves
-                // (via a `guid <> :guid` check), which means we need to "pretend" this is a new record,
-                // in order for `ensureValid` to actually throw `DUPLICATE_LOGIN`.
-                storage.value.ensureValid(login.copy(guid = null))
-                Result.CanBeCreated
+            val savedLoginByGuid = newLogin.guid?.let { storage.value.get(it) }
+            val savedLoginByElements = storage.value.getStoredLoginByImportantFields(newLogin)
+            // Check the merged login with the new password to display correct UI error states
+            val loginToCheck =
```

Comment:
I'm trying to understand what's going on here, and for such a small function`getNewLoginPersistenceAsync` is _very_ hard to reason about. As-is, I don't think we can land it - just imagine trying to understand this a few months from now! I've been reading and re-reading it, and I can't quite tell if it's correct.

You basically have three cases to deal with:

1) trivial stuff - malformed record, or a complete dupe of an existing record
2) entirely new record - guid, username+origin, etc, nothing matches any existing records
3) existing record being updated - either matched by guid, or by "the essence" of the login. Note that `username` is immutable for a record, so a username change is same as creating an entirely new record and covered by case (2).
   - Also: We may or may not know `newLogin.guid` - what does that depend on? When will we have a `guid` set on `newLogin` and when will we not?

Can you update this code in a way that makes it very clear which cases you're dealing with, and how?

Also: all of this logic should live inside of the `logins` rust library. It'll be easier to understand, and much faster since I bet you can just write it out in SQL. This code runs during user input, so we should try to make it quick!

Having all of this logic live in one place will ensure that we can actually easily understand what's going on, and aren't duplicating ourselves. For an example of conceptual duplication, take a look at `getStoredLoginByImportantFields` and the dupe check within `ensureValid` - https://github.com/mozilla/application-services/blob/master/components/logins/src/db.rs#L819.

I don't mind if we have a concise, clear and correct implementation of this logic here first, but we _really_ should move it down a layer.